### ISSUE TITLE
Add Check Support for Java 21 Pattern Matching for Switch Syntax: WhitespaceAround

### ISSUE THREAD

**Issue by mahfouz72** [checkstyle/checkstyle#15027]
child of https://github.com/checkstyle/checkstyle/issues/14961

I have read check documentation: https://checkstyle.org/checks/whitespace/whitespacearound.html#WhitespaceAround
I have downloaded the latest checkstyle from: https://checkstyle.org/cmdline.html#Download_and_Run
I have executed the cli and showed it below, as cli describes the problem better than 1,000 words

```bash
➜  src cat config.xml 
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
        "-//Puppy Crawl//DTD Check Configuration 1.3//EN"
        "http://www.puppycrawl.com/dtds/configuration_1_3.dtd">
<module name="Checker">
    <property name="charset" value="UTF-8"/>
    <module name="TreeWalker">
        <module name="WhitespaceAround"/>
    </module>
</module>
➜  src cat Test.java 
record ColoredPoint(int p, int x, String c) { }
record Rectangle(ColoredPoint upperLeft, ColoredPoint lowerRight) { }

public class Test {

    void test(Object obj) {
        int y = 0;
        if(y >= 2) { }      // violation
        switch (obj) {
            case ColoredPoint(int x, _, _) when(x >= 2) -> { }   // expected violation
            default -> { }
        }
    }

    void test2(Object obj) {
        int y=0; // 2 violations, no space before and after '='
        switch (obj) {
            case ColoredPoint(int x, _, _)when(x >= 2) -> { } // (expected) 2 violations, no space before and after 'when'
            case ColoredPoint(int x, _, _) when(x >= 0)  -> { }  // (expected) 1 violation, no space after 'when'
            default -> { }
        }
    }
}
➜  src javac --enable-preview --release 21 Test.java
Note: Test.java uses preview features of Java SE 21.
Note: Recompile with -Xlint:preview for details.
➜  src java -jar checkstyle-10.18.1-SNAPSHOT-all.jar -c config.xml Test.java
Starting audit...
[ERROR] /home/nick/IdeaProjects/tester/src/Test.java:8:9: 'if' is not followed by whitespace. [WhitespaceAround]
[ERROR] /home/nick/IdeaProjects/tester/src/Test.java:16:14: '=' is not followed by whitespace. [WhitespaceAround]
[ERROR] /home/nick/IdeaProjects/tester/src/Test.java:16:14: '=' is not preceded with whitespace. [WhitespaceAround]
Audit done.
Checkstyle ends with 3 errors.

```

---

Describe what you want in detail

violation on `when`.

**Comment by rnveach** [comment]
Token to add is `LITERAL_WHEN` and should be able to turn on/off.

Overlap already exists with WithspaceAfter, so I am good with this.

> case ColoredPoint(int x, _, _) when(x >= 2) -> { }

Add examples like:
````
case ColoredPoint(int x, _, _)when (x >= 2) -> { }
case ColoredPoint(int x, _, _)when(x >= 2) -> { }
````

Otherwise, I am good with this.

**Comment by mahfouz72** [comment]
```
record ColoredPoint(int p, int x, String c) { }
record Rectangle(ColoredPoint upperLeft, ColoredPoint lowerRight) { }

public class Test {

    void test(Object obj) {
        int y =0; // 2 violations, no space before and after '='
        switch (obj) {
            case ColoredPoint(int x, _, _)when(x >= 2) -> { } // (expected) 2 violations, no space before and after 'when'
            case ColoredPoint(int x, _, _) when(x >= 0)  -> { }  // (expected) 1 violation, no space before and after 'when'
            default -> { }
        }
    }
}
```


> Overlap already exists with WithspaceAfter, so I am good with this.

Yes, I consider this check as `WhitespaceAfter` and `WhitespaceBefore` simultaneously. we should support `LITERAL_WHEN` in both around and after

**Comment by rnveach** [comment]
For clarity, I meant to add the examples to the first post. We usually like to keep all details there so it is easier to go back to on what was agreed when the issue is approved.

**Comment by nrmancuso** [comment]
Approved, @mahfouz72 please extend issue description

**Comment by nrmancuso** [comment]
I have updated the example now that we have the `EXPR` token; there is no change in behavior. This is a false negative.

**Comment by rnveach** [comment]
Also waiting on IT updates
